src/actions/repeated-action.h

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright 2011 - 2020 José Expósito <jose.exposito89@gmail.com>
+ *
+ * This file is part of Touchégg.
+ *
+ * Touchégg is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License  as  published by  the  Free Software
+ * Foundation,  either version 3 of the License,  or (at your option)  any later
+ * version.
+ *
+ * Touchégg is distributed in the hope that it will be useful,  but  WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+ * A PARTICULAR PURPOSE.  See the  GNU General Public License  for more details.
+ *
+ * You should have received a copy of the  GNU General Public License along with
+ * Touchégg. If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef ACTIONS_REPEATED_ACTION_H_
+#define ACTIONS_REPEATED_ACTION_H_
+
+#include "actions/action.h"
+
+class RepeatedAction : public Action {
+ public:
+  using Action::Action;
+  virtual ~RepeatedAction() = default;
+
+  /**
+   * Execute gesture on begin if threshold passed
+   */
+  void onGestureBegin(const Gesture &gesture) override;
+  /**
+   * Execute gesture repeatedly
+   */
+  void onGestureUpdate(const Gesture &gesture) override;
+  /**
+   * Execute gesture on end if threshold passed
+   */
+  void onGestureEnd(const Gesture &gesture) override;
+
+  /**
+   * If your action requires some pre-setup, override this method. This is
+   * executed unconditionally upon gesture beginning
+   */
+  virtual void executePrelude();
+  /**
+   * If your action requires some post-setup, override this method. This is
+   * executed unconditionally upon gesture ending
+   */
+  virtual void executePostlude();
+  /**
+   * Override this method to configure what is executed for succesful gesture
+   */
+  virtual void executeAction(const Gesture &gesture) = 0;
+  /**
+   * Override this method to configure what is executed for reversed gesture
+   */
+  virtual void executeReverse(const Gesture &gesture) = 0;
+
+ protected:
+  bool repeat = false;
+  int repeatPercentage = 0;
+  bool onBegin = true;
+  /** Safeguard against executing onBegin action multiple times when execution
+   * threshold is surpassed multiple times */
+  bool onBeginExecuted = false;
```

Comment:
The method `onGestureBegin(const Gesture &gesture)` is guaranteed to be called once, so I think it is safe to remove this flag.

Reply:
You see, I made it so that `onGestureBegin` doesn't always execute the gesture, it only exectues it if the threshold was passed during the beginning. If the threshold wasn't passed during the beginning, we execute the action during `onGestureUpdate` the first time threshold is passed.
Now that I think about it, this is not the same behaviour as it was before. I think it might be a good idea to keep the old behaviour, but also to add this new behaviour with a new option `<on>threshold</on>`

Reply:
Yes, when the `on` tag is set to `begin`, the action should be executed unconditionally at the beginning of the gesture so actions feel quick to perform.
I'd rather not include any extra options because the config file is already pretty complex. In addition, I'd have to add it and maintain it in the (work in progress) GUI.

Reply:
fix: 86170b44b507cc72077c369d560298a83c57fc2b